codec/enforceproto/extraneous_fields.go

```diff
@@ -0,0 +1,210 @@
+var typeToIndexFieldsForType = make(map[reflect.Type]map[protowire.Number]*tagAndProtoType)
+
+func indexFieldsForType(typ reflect.Type) (map[protowire.Number]*tagAndProtoType, error) {
+	typeToIndexFieldsMu.Lock()
+	defer typeToIndexFieldsMu.Unlock()
+
+	index, ok := typeToIndexFieldsForType[typ]
+	if ok {
+		return index, nil
+	}
+
+	index = make(map[protowire.Number]*tagAndProtoType)
+	for i, n := 0, typ.NumField(); i < n; i++ {
+		field := typ.Field(i)
+		protoTag, ok := field.Tag.Lookup("protobuf")
+		if !ok {
+			// Try if it is a protobuf_oneof
+			protoTag, ok = field.Tag.Lookup("protobuf_oneof")
+			if !ok {
+				continue
+			}
+		}
+		byCommaSplits := strings.Split(protoTag, ",")
+		if len(byCommaSplits) == 0 {
+			// Strange, how can a protobuf tag not be populated?
+			continue
+		}
+		wireType, protoFieldNumStr := byCommaSplits[0], byCommaSplits[1]
+		protoField, err := strconv.ParseUint(protoFieldNumStr, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("invalid protoFieldNum in %q", protoFieldNumStr)
+		}
+		tagNum := protowire.Number(protoField)
+
+		wireTypeType, ok := wireTypeStrToWireType[wireType]
+		if !ok {
+			panic(fmt.Sprintf("unhandled wireType %q", wireType))
+		}
+		index[tagNum] = &tagAndProtoType{
+			tagNum:    tagNum,
+			protoType: wireType,
+			wireType:  wireTypeType,
+		}
+	}
+
+	// By this point we should memoize the index.
+	typeToIndexFieldsForType[typ] = index
+
+	return index, nil
+}
+
+var wireTypeStrToWireType = map[string]protowire.Type{
+	"varint":      0,
+	"fixed64":     1,
+	"bytes":       2,
+	"start_group": 3,
+	"end_group":   4,
+	"fixed32":     5,
+}
+
+func wireTypeToString(wt protowire.Type) string {
+	switch wt {
+	case 0:
+		return "varint"
+	case 1:
+		return "fixed64"
+	case 2:
+		return "bytes"
+	case 3:
+		return "start_group"
+	case 4:
+		return "end_group"
+	case 5:
+		return "fixed32"
+	default:
+		return fmt.Sprintf("Unknown type: %d", wt)
+	}
+}
+
+type MismatchedWireType struct {
+	GotWireType  protowire.Type
+	WantWireType protowire.Type
+	TagNum       protowire.Number
+}
+
+func (mwt *MismatchedWireType) String() string {
+	return fmt.Sprintf("Mismatched{TagNum: %d, GotWireType: %q != WantWireType: %q}",
+		mwt.TagNum, wireTypeToString(mwt.GotWireType), wireTypeToString(mwt.WantWireType))
+}
+
+func (mwt *MismatchedWireType) Error() string {
+	return mwt.String()
+}
+
+var _ error = (*MismatchedWireType)(nil)
+
+// CheckMismatchedProtoFields walks through the protobuf serialized bytes in b, and tries to
+// compare field numbers and wireTypes against what msg expects. The error returned if non-nil will contain
+// the listing of the extraneous fields by tagNumber and wireType, as well as mismatched
+// wireTypes.
+func CheckMismatchedProtoFields(b []byte, msg gogoproto.Message) error {
+	if msg == nil {
+		return errors.New("expecting a non-nil message")
+	}
+	rv := reflect.ValueOf(msg)
+	rt := rv.Elem().Type()
+
+	index, err := indexFieldsForType(rt)
```

Comment:
This should get cached in a dictionary of types so that reflection only needs to happen the first type a struct type is encountered. This likely means that `CheckMismatchedProtoFields` needs to be the receiver on some struct that holds a cache.

Reply:
It is cached, that function indexFieldsForType has a cache inside it
https://github.com/cosmos/cosmos-sdk/blob/ad3dba5731f936d6c51fdcdbcefdd63aa4366343/codec/enforceproto/extraneous_fields.go#L30-L77

Reply:
I now see what you mean, sorry brain freeze and slowly getting thawed, changes coming up slowly :)

Reply:
I implemented that in a follow-on commit.